Added support for mouseEnter and mouseLeave events on a row

Techn1x opened the pull request:
I needed to be able to detect hover events on a row, so I implemented this. It seems to work and I think it's correct - let me know if any further changes are required.

buschtoens commented:
> It seems to work
> let me know if any further changes are required.

Please add tests asserting that :smile:

Techn1x commented:
whoops, thanks. As far as I can tell I added the tests correctly, but Travis is saying `Expected 4 assertions, but 2 were run` for that test, and I am not sure why it's not triggering all 4 assertions..

buschtoens commented on tests/integration/components/lt-body-test.js:
```diff
@@ -148,16 +148,20 @@ test('row expansion - multiple', function(assert) {
 });
 
 test('row actions', function(assert) {
```
```js
test('row actions', async function(assert) {
```

By turning this into an [`async` function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function), you can use the `await` keyboard in the function body. [ember-native-dom-helpers](https://github.com/cibernox/ember-native-dom-helpers#acceptance-tests) returns a promise from all the interaction helpers, that resolves when the world has settled and there are no pending renders, requests or updates.

buschtoens commented on tests/integration/components/lt-body-test.js:
```diff
@@ -148,16 +148,20 @@ test('row expansion - multiple', function(assert) {
-  this.render(hbs `{{lt-body table=table sharedOptions=sharedOptions onRowClick=(action 'onRowClick') onRowDoubleClick=(action 'onRowDoubleClick')}}`);
+  this.on('onRowMouseEnter', (row) => assert.ok(row));
+  this.on('onRowMouseLeave', (row) => assert.ok(row));
+  this.render(hbs `{{lt-body table=table sharedOptions=sharedOptions onRowClick=(action 'onRowClick') onRowDoubleClick=(action 'onRowDoubleClick') onRowMouseEnter=(action 'onRowMouseEnter') onRowMouseLeave=(action 'onRowMouseLeave')}}`);
```
```js
await this.render(/* ... */);
```

Use `await` here please. It does not have any effect yet with the current `ember-qunit` version that's installed on your branch, but when https://github.com/offirgolan/ember-light-table/pull/531 is merged, it will. Please refer to [RFC 232](https://github.com/emberjs/rfcs/blob/master/text/0232-simplify-qunit-testing-api.md#changes-from-current-system) for more info.

buschtoens commented on tests/integration/components/lt-body-test.js:
```diff
@@ -148,16 +148,20 @@ test('row expansion - multiple', function(assert) {
   click(row);
   triggerEvent(row, 'dblclick');
+  triggerEvent(row, 'mouseenter');
+  triggerEvent(row, 'mouseleave');
```
And then lastly use `await` for all interaction helpers here as well. This will make sure that the test runner waits until the world has settled and all expected assertions are actually run.

buschtoens commented on addon/components/lt-body.js:
```diff
@@ -462,6 +462,26 @@ export default Component.extend({
+     * onRowMouseEnter action.
+     * @event onRowMouseEnter
+     * @param  {Row}   row The row that was entered
+     * @param  {Event}   event   The click event
```
*mouse event

buschtoens commented on addon/components/lt-body.js:
```diff
@@ -462,6 +462,26 @@ export default Component.extend({
+     * @param  {Row}   row The row that was entered
+     * @param  {Event}   event   The click event
+     */
+    onRowMouseEnter(/* row */) {
```
```js
onRowMouseEnter(/* row, event */) {
```

buschtoens commented on addon/components/lt-body.js:
```diff
@@ -462,6 +462,26 @@ export default Component.extend({
+     * onRowMouseLeave action.
+     * @event onRowMouseLeave
+     * @param  {Row}   row The row that was left
+     * @param  {Event}   event   The click event
```
*mouse event

buschtoens commented on addon/components/lt-body.js:
```diff
@@ -462,6 +462,26 @@ export default Component.extend({
+     * @param  {Row}   row The row that was left
+     * @param  {Event}   event   The click event
+     */
+    onRowMouseLeave(/* row */) {
```
```js
onRowMouseEnter(/* row, event */) {
```

buschtoens commented:
Hm. I pulled your branch and merged it with #531, but still there are only two assertions run...

Your code works correctly though. It seems to be a problem with `triggerEvent(row, 'mouseenter')` and `triggerEvent(row, 'mouseleave')`...

Techn1x commented:
No problem at all, I am happy to keep using the PR for my project until you find some time! Just wanted to make sure there wasn't any blockers to this going ahead.

Good luck with the new job!

buschtoens commented:
@Techn1x Can you please rebase and make sure that all tests pass? Then we can finally merge! :tada: :smile:

Techn1x commented:
Rebased, and ensured the tests were in the same format/style as your modernised test updates

Unfortunately, 2 of the 4 assertions are still not running and I'm not sure why. `mouseenter` and `mouseleave` are supported events for `ember-native-dom-helpers`

Techn1x commented:
@buschtoens Sorry for leaving this for so long, I've rebased (again) and fixed up the test to pass completely. Should be good to merge.

For whatever reason, the 'mouseenter' and 'mouseleave' events weren't triggering unless there were also 'mouseover' and 'mouseout' events. Does this sound right to you?

Techn1x commented:
@snewcomer have you got time to merge this?

Techn1x commented:
Reopened here: https://github.com/offirgolan/ember-light-table/pull/685
